bios/stage-4/src/main.rs

```diff
@@ -0,0 +1,222 @@
+    }
+
+    log::info!("BIOS boot");
+
+    let page_tables = create_page_tables(&mut frame_allocator);
+
+    let kernel_slice = {
+        let ptr = kernel_start.as_u64() as *const u8;
+        unsafe { slice::from_raw_parts(ptr, usize_from(kernel_size)) }
+    };
+    let kernel = Kernel::parse(kernel_slice);
+
+    let system_info = SystemInfo {
+        framebuffer_addr: PhysAddr::new(info.framebuffer.region.start),
+        framebuffer_info,
+        rsdp_addr: detect_rsdp(),
+    };
+
+    load_and_switch_to_kernel(kernel, frame_allocator, page_tables, system_info);
+}
+
+fn init_logger(info: BiosFramebufferInfo) -> FrameBufferInfo {
+    let framebuffer_info = FrameBufferInfo {
+        byte_len: info.region.len.try_into().unwrap(),
+        width: info.width.into(),
+        height: info.height.into(),
+        pixel_format: match info.pixel_format {
+            bootloader_x86_64_bios_common::PixelFormat::Rgb => PixelFormat::Rgb,
+            bootloader_x86_64_bios_common::PixelFormat::Bgr => PixelFormat::Bgr,
+            bootloader_x86_64_bios_common::PixelFormat::Unknown {
+                red_position,
+                green_position,
+                blue_position,
+            } => PixelFormat::Unknown {
+                red_position,
+                green_position,
+                blue_position,
+            },
+        },
+        bytes_per_pixel: info.bytes_per_pixel.into(),
+        stride: info.stride.into(),
+    };
+
+    let framebuffer = unsafe {
+        core::slice::from_raw_parts_mut(
+            info.region.start as *mut u8,
+            info.region.len.try_into().unwrap(),
+        )
+    };
+
+    bootloader_x86_64_common::init_logger(framebuffer, framebuffer_info);
+
+    framebuffer_info
+}
+
+/// Creates page table abstraction types for both the bootloader and kernel page tables.
+fn create_page_tables(frame_allocator: &mut impl FrameAllocator<Size4KiB>) -> PageTables {
+    // We identity-mapped all memory, so the offset between physical and virtual addresses is 0
+    let phys_offset = VirtAddr::new(0);
+
+    // copy the currently active level 4 page table, because it might be read-only
+    let bootloader_page_table = {
+        let frame = x86_64::registers::control::Cr3::read().0;
+        let table: *mut PageTable = (phys_offset + frame.start_address().as_u64()).as_mut_ptr();
+        unsafe { OffsetPageTable::new(&mut *table, phys_offset) }
+    };
+
+    // create a new page table hierarchy for the kernel
+    let (kernel_page_table, kernel_level_4_frame) = {
+        // get an unused frame for new level 4 page table
+        let frame: PhysFrame = frame_allocator.allocate_frame().expect("no unused frames");
+        log::info!("New page table at: {:#?}", &frame);
+        // get the corresponding virtual address
+        let addr = phys_offset + frame.start_address().as_u64();
+        // initialize a new page table
+        let ptr = addr.as_mut_ptr();
+        unsafe { *ptr = PageTable::new() };
```

Comment:
We might want to use `core::ptr::write` here instead. I guess that depends on whether we regard the memory that was just allocated as uninitialized. I'd argue that it is uninitialized, though I don't think this will cause miscompilations considering that this is somewhat magic memory from outside the AM.

Reply:
Changed in https://github.com/rust-osdev/bootloader/pull/232/commits/26a6aee6ef5311b7dd4882eb2ae007f1c2a7a62c